src/modules/application/controllers/controllerCreate.js

```diff
@@ -1,26 +1,45 @@
 import { databasePrisma } from "../../../prismaClient.js";
 
-export const handleCreate = async function (req, res) {
-  const data = req.body;
+export const handleCreate = async function (req) {
+  const { applicantId, listingId, coverLetter, companyId } = req.body;
+
+  const hasUserCreatedApplication = await databasePrisma.application.count({
+    where: {
+      applicantId,
+      listingId,
+    },
+  });
+
+  if (hasUserCreatedApplication > 0) {
+    return Promise.resolve({
+      status: 409,
+      message: "You've already created an application on this listing",
+    });
+  }
+
```

Comment:
Rather than a count we should be getting the listing itself to see if it exist, and then checking if it has an existing application attached to it matching the applicant's id.  At the moment I can put in any id for the listing and get a 400 error `"message": "application ID, listing ID, cover letter, and company ID are required"`, when I should be told the listing doesn't exist. Same applies to the company id, but I think that shouldn't be a requirement in the request we should just get the id from the listing itself when we check it exists.

We should also be checking a user has the Applicant role, but I think that should be done in verification middleware before this point.

Reply:
Fixed